tests/e2e/e2e_setup_test.go

```diff
@@ -0,0 +1,315 @@
+)
+
+type IntegrationTestSuite struct {
+	suite.Suite
+
+	tmpDirs      []string
+	chain        *chain
+	dkrPool      *dockertest.Pool
+	dkrNet       *dockertest.Network
+	valResources map[string][]*dockertest.Resource
+}
+
+func TestIntegrationTestSuite(t *testing.T) {
+	suite.Run(t, new(IntegrationTestSuite))
+}
+
+func (s *IntegrationTestSuite) SetupSuite() {
+	s.T().Log("setting up e2e integration test suite...")
+
+	var err error
+	s.chain, err = newChain()
+	s.Require().NoError(err)
+
+	s.dkrPool, err = dockertest.NewPool("")
+	s.Require().NoError(err)
+
+	s.dkrNet, err = s.dkrPool.CreateNetwork(fmt.Sprintf("%s-testnet", s.chain.id))
+	s.Require().NoError(err)
+
+	s.valResources = make(map[string][]*dockertest.Resource)
+
+	// The boostrapping phase is as follows:
+	//
+	// 1. Initialize Osmosis validator nodes.
+	// 2. Create and initialize Osmosis validator genesis files (one chain)
+
+	s.T().Logf("starting e2e infrastructure for chain A; chain-id: %s; datadir: %s", s.chain.id, s.chain.dataDir)
+	s.initNodes(s.chain)
+	s.initGenesis(s.chain)
+	s.initValidatorConfigs(s.chain)
+	s.runValidators(s.chain, 0)
+}
+
+func (s *IntegrationTestSuite) TearDownSuite() {
+	if str := os.Getenv("OSMOSIS_E2E_SKIP_CLEANUP"); len(str) > 0 {
+		skipCleanup, err := strconv.ParseBool(str)
+		s.Require().NoError(err)
+
+		if skipCleanup {
+			return
+		}
+	}
+
+	s.T().Log("tearing down e2e integration test suite...")
+
+	for _, vr := range s.valResources {
+		for _, r := range vr {
+			s.Require().NoError(s.dkrPool.Purge(r))
+		}
+	}
+
+	s.Require().NoError(s.dkrPool.RemoveNetwork(s.dkrNet))
+
+	os.RemoveAll(s.chain.dataDir)
+
+	for _, td := range s.tmpDirs {
+		os.RemoveAll(td)
+	}
+}
+
+func (s *IntegrationTestSuite) initNodes(c *chain) {
+	s.Require().NoError(c.createAndInitValidators(2))
+
+	// initialize a genesis file for the first validator
+	val0ConfigDir := c.validators[0].configDir()
+	for _, val := range c.validators {
+		s.Require().NoError(
+			addGenesisAccount(val0ConfigDir, "", initBalanceStr, val.keyInfo.GetAddress()),
+		)
+	}
+
+	// copy the genesis file to the remaining validators
+	for _, val := range c.validators[1:] {
+		_, err := copyFile(
+			filepath.Join(val0ConfigDir, "config", "genesis.json"),
+			filepath.Join(val.configDir(), "config", "genesis.json"),
+		)
+		s.Require().NoError(err)
+	}
+}
+
+func (s *IntegrationTestSuite) initGenesis(c *chain) {
+	serverCtx := server.NewDefaultContext()
+	config := serverCtx.Config
+
+	config.SetRoot(c.validators[0].configDir())
+	config.Moniker = c.validators[0].moniker
+
+	genFilePath := config.GenesisFile()
+	appGenState, genDoc, err := genutiltypes.GenesisStateFromGenFile(genFilePath)
+	s.Require().NoError(err)
+
+	var bankGenState banktypes.GenesisState
+	s.Require().NoError(cdc.UnmarshalJSON(appGenState[banktypes.ModuleName], &bankGenState))
+
+	bankGenState.DenomMetadata = append(bankGenState.DenomMetadata, banktypes.Metadata{
+		Description: "An example stable token",
```

Comment:
Oh wow, I didn't know there was a description field available in the bank module